aten/src/ATen/native/quantized/cpu/qadd.cpp

```diff
@@ -43,23 +43,59 @@ Tensor _add_out(Tensor& out, const Tensor& self, const Tensor& other) {
 
 template <bool ReLUFused = false>
 Tensor _add_scalar_out(Tensor& out, const Tensor& self, Scalar other) {
-  int64_t zero_point = out.q_zero_point();
-  double scale = out.q_scale();
-  int64_t self_zero_point = self.q_zero_point();
-  double self_scale = self.q_scale();
-
-  auto iter = TensorIterator::unary_op(out, self);
-  AT_DISPATCH_QINT_TYPES(out.scalar_type(), "qadd", [&]() {
-    cpu_kernel(iter, [&](scalar_t a) -> scalar_t {
-      const auto da = at::dequantize_val(self_scale, self_zero_point, a);
-      double c = da + other.toFloat();
-      auto quant_val = at::quantize_val<scalar_t>(scale, zero_point, c);
-      auto dequant_val = at::dequantize_val(scale, zero_point, quant_val);
-      if (ReLUFused) {
-        c = std::max<float>(c, 0.0);
-      }
-      return at::quantize_val<scalar_t>(scale, zero_point, c);
-    });
+  TORCH_CHECK(self.qscheme() == kPerTensorAffine,
+              "Only per tensor affine is supported for now!!");
+  // To implement tensor-scalar addition in quantized space, we simply
+  // adjust the quantization parameters based on the following rules:
+  //
+  // Let s = scale, z = zero point, c = other.toFloat(),
+  //     q_min = lowest representable value of scalar type
+  //     q_max = highest representable value of scalar type
+  //
+  //  Let s' = the calculated scale or the output
+  //      z' = the calculated zero-point for the output
+  //
+  //  If q_min > z - round(c / s)
+  //    s' = [(q_max - z)*s + c]/[q_max - q_min]
+  //    z' = q_min
+  //
+  //  If q_max < z - round(c / s)
+  //    s' = -[(q_min - z)*s + c]/[q_max - q_min]
+  //    z' = q_max
+  //
+  //  Else
+  //    s' = s
+  //    z = z - round(c / s)
```

Comment:
The part that is missing is requantizing  the quantized values when the addition is out of bound:
We need to do:
 // Let s = scale, z = zero point, c = other.toFloat(),    c_q = round(c/s)
  //     q_min = lowest representable value of scalar type
  //     q_max = highest representable value of scalar type
  //
  //  Let s' = the calculated scale or the output
  //      z' = the calculated zero-point for the output
  If q_min > c_q
  //    s' = [(q_max - (z -  c_q)]/[q_max - q_min] *s
  //    z' = q_min
  //    Xq'  = torch.quantize_linear(Xq.dequantize() + c_q.dequantize() , s', z')
  //  If q_max < z - c_q
  //    s' = [z - c_q -q_min]/[q_max - q_min] *s
  //    z' = q_max
   //    Xq'  = torch.quantize_linear(Xq.dequantize() + c_q.dequantize(), s', z')

  //  Else
  //    s' = s
  //    z' = z - c_q

This op is identical in floating point semantics to:

 y = quantize( ( x_q.dequantize()+c_q.dequantize(), s',z') ).dequantize()